plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/index.tsx

```diff
@@ -31,49 +33,59 @@ const Edit = ( props: ProductCollectionEditComponentProps ) => {
 		[ clientId ]
 	);
 
-	const productCollectionUIStateInEditor =
-		getProductCollectionUIStateInEditor( {
-			hasInnerBlocks,
+	const { productCollectionUIStateInEditor, isLoading } =
+		useProductCollectionUIState( {
 			location,
-			attributes: props.attributes,
+			attributes,
+			hasInnerBlocks,
 			usesReference: props.usesReference,
 		} );
 
-	/**
-	 * Component to render based on the UI state.
-	 */
-	let Component,
-		isUsingReferencePreviewMode = false;
-	switch ( productCollectionUIStateInEditor ) {
-		case ProductCollectionUIStatesInEditor.COLLECTION_PICKER:
-			Component = ProductCollectionPlaceholder;
-			break;
-		case ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER:
-			Component = ProductPicker;
-			break;
-		case ProductCollectionUIStatesInEditor.VALID:
-			Component = ProductCollectionContent;
-			break;
-		case ProductCollectionUIStatesInEditor.VALID_WITH_PREVIEW:
-			Component = ProductCollectionContent;
-			isUsingReferencePreviewMode = true;
-			break;
-		default:
-			// By default showing collection chooser.
-			Component = ProductCollectionPlaceholder;
+	// Show spinner while calculating Editor UI state.
+	if ( isLoading ) {
+		return (
+			<Flex justify="center" align="center">
+				<Spinner />
+			</Flex>
+		);
 	}
 
+	const productCollectionContentProps: ProductCollectionContentProps = {
+		...props,
+		openCollectionSelectionModal: () => setIsSelectionModalOpen( true ),
+		location,
+		isUsingReferencePreviewMode:
+			productCollectionUIStateInEditor ===
+			ProductCollectionUIStatesInEditor.VALID_WITH_PREVIEW,
+	};
+
+	const renderComponent = () => {
+		switch ( productCollectionUIStateInEditor ) {
+			case ProductCollectionUIStatesInEditor.COLLECTION_PICKER:
+				return <ProductCollectionPlaceholder { ...props } />;
+			case ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER:
+			case ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE:
+				return (
+					<ProductPicker
+						{ ...props }
+						isDeletedProductReference={ true }
+					/>
+				);
```

Comment:
PRODUCT_REFERENCE_PICKER case requires a separate return without `isDeletedProductReference`. Because in the current shape we display "Previously selected product is no longer available." message even if adding new collection with no previous reference.

```suggestion
			case ProductCollectionUIStatesInEditor.PRODUCT_REFERENCE_PICKER:
			    return <ProductPicker { ...props } />;
			case ProductCollectionUIStatesInEditor.DELETED_PRODUCT_REFERENCE:
				return (
					<ProductPicker
						{ ...props }
						isDeletedProductReference={ true }
					/>
				);
```

Reply:
Aha! Good catch. Fixed it in 9f04f11